### PR TITLE
chore: Add case-insensitive string set utility

### PR DESCRIPTION
**Pull request by nilo19**
A new structure called IgnoreCaseSet has been introduced into the pkg/util/sets package. This constructs a set of strings that is case-insensitive, providing methods for applying standard set operations. It's beneficial for handling inputs where case sensitivity isn't important or might cause issues.

#### What this PR does / why we need it:

chore: Add case-insensitive string set utility

A new structure called IgnoreCaseSet has been introduced into the pkg/util/sets package. This constructs a set of strings that is case-insensitive, providing methods for applying standard set operations. It's beneficial for handling inputs where case sensitivity isn't important or might cause issues.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/provider/azure.go [hunk 1]
```diff
@@ -48,7 +48,6 @@ import (
 	cloudnodeutil "k8s.io/cloud-provider/node/helpers"
 	nodeutil "k8s.io/component-helpers/node/util"
 	"k8s.io/klog/v2"
-	"sigs.k8s.io/yaml"
 
 	"sigs.k8s.io/cloud-provider-azure/pkg/azclient"
```
Comment (nilo19) [inline pkg/provider/azure.go:52]: Looks like there is an issue with goimports checker

File: pkg/util/sets/string.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+// Insert adds the given items to the set. It only works if the set is initialized.
+func (s *IgnoreCaseSet) Insert(items ...string) {
+	var lowerItems []string
+	for _, item := range items {
+		lowerItems = append(lowerItems, strings.ToLower(item))
+	}
+	for _, item := range lowerItems {
+		s.set.Insert(strings.ToLower(item))
```
Comment (lodrem) [inline pkg/util/sets/string.go:47]: Why call `strings.ToLower` twice?

File: pkg/util/sets/string.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+func (s *IgnoreCaseSet) Has(item string) bool {
+	if !s.Initialized() {
+		return false
+	}
+	return s.set.Has(strings.ToLower(item))
+}
+
+// Initialized returns true if the set is initialized.
+func (s *IgnoreCaseSet) Initialized() bool {
+	return s != nil && s.set != nil
```
Comment (lodrem) [inline pkg/util/sets/string.go:83]: When would `s == nil` or `s.set == nil`? Shouldn't the caller always init the set using `NewString`?

Comment (lodrem) [reply pkg/util/sets/string.go:83]: need to do defensive checking for all methods right now...

File: pkg/provider/azure.go (hunk 1 above)
Comment (lzhecheng) [reply pkg/provider/azure.go:52]: If this pkg is not moved, will there be a failure?

Comment (nilo19) [reply pkg/provider/azure.go:52]: I don't think it should be here if not used.

File: pkg/util/sets/string.go (hunk 2 above)
Comment (nilo19) [reply pkg/util/sets/string.go:47]: That is a mistake, the second one is redundant.

File: pkg/util/sets/string.go (hunk 3 above)
Comment (nilo19) [reply pkg/util/sets/string.go:83]: > When would `s == nil` or `s.set == nil`? Shouldn't the caller always init the set using `NewString`?

unit test

Comment (nilo19) [reply pkg/util/sets/string.go:83]: > need to do defensive checking for all methods right now...

I check it in various methods like delete and safe insert.

Comment (lodrem) [reply pkg/util/sets/string.go:83]: It’s confusing to decide whether to use `sets.insert` or `SafeInsert`. I suggest maintaining the same interface as `sets.Set[string]`.

> unit test

Could we use the constructor to init the variable or data even in unit tests?

Comment (nilo19) [reply pkg/util/sets/string.go:83]: This is better than sets.Set[string], as it has no nil checker in Insert as well. Initialized() is used within the new Set but I expose it to offer more flexibility in development.

File: pkg/provider/azure.go (hunk 1 above)
Comment (nilo19) [reply pkg/provider/azure.go:52]: fix: sigs.k8s.io/yaml is not local module so it should not be grouped with local ones.

**Review by feiskyer — APPROVED** [review]
/lgtm

**Pull request merged**